chalk-engine/src/logic.rs

```diff
@@ -412,67 +378,90 @@ impl<C: Context> Forest<C> {
         depth: StackIndex,
         mut strand: Strand<C>,
     ) -> StrandResult<C, ()> {
-        info_heading!(
-            "pursue_strand(table={:?}, depth={:?}, ex_clause={:#?}, selected_subgoal={:?})",
-            self.stack[depth].table,
-            depth,
-            strand.infer.debug_ex_clause(&strand.ex_clause),
-            strand.selected_subgoal,
-        );
-
-        // If no subgoal has yet been selected, select one.
-        while strand.selected_subgoal.is_none() {
-            if strand.ex_clause.subgoals.len() == 0 {
-                if strand.ex_clause.floundered_subgoals.is_empty() {
-                    return self.pursue_answer(depth, strand);
-                }
-
-                self.reconsider_floundered_subgoals(&mut strand.ex_clause);
+        let mut allow_coinduction = false;
+        loop {
+            info_heading!(
+                "pursue_strand(table={:?}, depth={:?}, ex_clause={:#?}, selected_subgoal={:?})",
+                self.stack[depth].table,
+                depth,
+                strand.infer.debug_ex_clause(&strand.ex_clause),
+                strand.selected_subgoal,
+            );
 
+            // If no subgoal has yet been selected, select one.
+            while strand.selected_subgoal.is_none() {
                 if strand.ex_clause.subgoals.is_empty() {
-                    assert!(!strand.ex_clause.floundered_subgoals.is_empty());
-                    return Err(StrandFail::Floundered);
-                }
+                    if !strand.ex_clause.floundered_subgoals.is_empty() {
+                        self.reconsider_floundered_subgoals(&mut strand.ex_clause);
+                        if strand.ex_clause.subgoals.is_empty() {
+                            assert!(!strand.ex_clause.floundered_subgoals.is_empty());
+                            return Err(StrandFail::Floundered);
+                        }
+                        continue;
+                    }
+                    if !strand.ex_clause.delayed_coinductive_cycles.is_empty() {
+                        strand
+                            .ex_clause
+                            .subgoals
+                            .extend(strand.ex_clause.delayed_coinductive_cycles.drain(..));
+                        allow_coinduction = true;
```

Comment:
Hmm, I have to think about if that is possible. I don't think so, because the `PositiveCycle` is returned before any chance to make a new subgoal is reached.

Reply:
I also suspect not, but I wanted to trace out the flow and convince myself. :)

Reply:
FWIW, I'll add a comment to make sure this assertion explicit. (And also go through and verify).

Reply:
I also don't know what it would mean to *add* subgoals to a strand? Sure, we move subgoals to floundering and to delayed_coinductive, but they aren't "created" really after the initial `simplify_hh_goal`. Otherwise, if new subgoals can be added at any time, then that makes keeping track of whether or not a strand is "complete" much more difficult.

Reply:
They can in fact be created by every unification. This is partly what I've been attempting to refactor away, but I don't think we're going to be able to do that completely.

Reply:
Although I realize now that we actually *could*